pkg/collector/pipeline_test.go

```diff
@@ -121,3 +123,89 @@ func TestDumpReports(t *testing.T) {
 		})
 	}
 }
+
+// Custom annotations
+
+var clientCountries = map[string]string{
+	"192.0.2.1": "US",
+	"192.0.2.2": "CA",
+}
+
+var serverZones = map[string]string{
+	"203.0.113.75": "us-east1-a",
+	"203.0.113.76": "us-west1-b",
+}
+
+type geoAnnotator struct{}
+
+func (g geoAnnotator) ProcessReports(batch *ReportBatch) {
+	batch.Annotation = clientCountries[batch.ClientIP]
+	for i, _ := range batch.Reports {
+		batch.Reports[i].Annotation = serverZones[batch.Reports[i].ServerIP]
+	}
+}
+
+type stashReports struct {
```

Comment:
Thinking aloud: Could you get away with: `type stashReports = *ReportBatch`?

Reply:
I tried that; you can't attach a method to a pointer type.